core/src/main/java/io/questdb/cairo/wal/seq/SeqTxnTracker.java

```diff
@@ -25,23 +25,39 @@
 package io.questdb.cairo.wal.seq;
 
 import io.questdb.cairo.ErrorTag;
+import io.questdb.cairo.wal.O3JobParallelismRegulator;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.Rnd;
 import io.questdb.std.Unsafe;
+import io.questdb.std.datetime.microtime.MicrosecondClockImpl;
 import org.jetbrains.annotations.TestOnly;
 
-public class SeqTxnTracker {
+public class SeqTxnTracker implements O3JobParallelismRegulator {
+    private static final Log LOG = LogFactory.getLog(SeqTxnTracker.class);
     private static final long SEQ_TXN_OFFSET = Unsafe.getFieldOffset(SeqTxnTracker.class, "seqTxn");
     private static final long SUSPENDED_STATE_OFFSET = Unsafe.getFieldOffset(SeqTxnTracker.class, "suspendedState");
     private static final long WRITER_TXN_OFFSET = Unsafe.getFieldOffset(SeqTxnTracker.class, "writerTxn");
+    private final Rnd rnd;
     private volatile String errorMessage = "";
     private volatile ErrorTag errorTag = ErrorTag.NONE;
+    private int maxRecordedInflightPartitions = 1;
+    // positive int: holds max parallelism
+    // negative int: holds backoff counter
+    private int regulationValue = Integer.MAX_VALUE;
     @SuppressWarnings("FieldMayBeFinal")
     private volatile long seqTxn = -1;
     // -1 suspended
     // 0 unknown
     // 1 not suspended
     private volatile int suspendedState = 0;
+    private long walBackoffUntil = -1;
     private volatile long writerTxn = -1;
 
+    public SeqTxnTracker(Rnd rnd) {
```

Comment:
`rnd` parameter is not needed. Code never passes anything but `null` or `new Rnd()`.